src/lib/Parser.hs

```diff
@@ -341,34 +359,41 @@ interfaceInstance = do
   (p, pos) <- withPos letPat
   ann <- annot uType
   keyWord WhereKW
-  rhs <- mkRhs ann <$> (withSrc $ URecord <$> interfaceRecordFields "=")
-  let (ann', rhs') = addImplicitImplicitArgs pos (Just ann) rhs
-  return $ ULet InstanceLet (p, ann') rhs'
+  rhsEither <- mkConstructorCall ann <$> (withSrc $ (URecord . NoExt) <$>
+      interfaceRecordFields "=")
+  case rhsEither of
+    Left err  -> fail err
+    Right rhs -> let (ann', rhs') = addImplicitImplicitArgs pos (Just ann) rhs
+                 in return $ ULet InstanceLet (p, ann') rhs'
   where
     -- Here, we are traversing the type annotation to retrieve the name of
     -- the interface and generate its corresponding constructor.
-    mkRhs (WithSrc _ (UPi _ _ lrhs))      rhs = mkRhs lrhs rhs
-    mkRhs (WithSrc _ (UApp _ (WithSrc _ (UVar v)) _)) rhs =
-      (var . nameToStr . mkInterfaceConsName . varName) v `mkApp` rhs
-    -- We let this case through to fail during type checking.
-    mkRhs _ rhs = rhs
+    mkConstructorCall (WithSrc _ (UPi _ arr typ)) record
+      | arr `elem` [ClassArrow, ImplicitArrow] = mkConstructorCall typ record
+    -- The innermost UApp is an application of a variable that has the name of
+    -- the interface to a type. We retrieve its name and derive the name of
+    -- its automatically generated constructor, which we apply to the record
+    -- passed as a parameter.
+    mkConstructorCall (WithSrc _ (UApp _ (WithSrc _ (UVar v)) _)) record =
+      Right $ (var . nameToStr . mkInterfaceConsName . varName) v `mkApp` record
+    -- Given the instance "I Int Float" of a multi-parameter interface I, we
+    -- end up with something similar to UApp _ (UApp _ I Int) Float. The below
+    -- traverses UApps to get to the innermost one..
+    mkConstructorCall (WithSrc _ (UApp _ func _)) record =
+      mkConstructorCall func record
```

Comment:
I feel a little uneasy about this, because how do we know that someone will use a variable and not e.g. `..i` which will probably parse as a valid type. This would deserve a different error message than the one below.